lightning/src/ln/channelmanager.rs

```diff
@@ -3096,11 +3146,12 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		{
 			let mut channel_state_lock = self.channel_state.lock().unwrap();
 			let channel_state = &mut *channel_state_lock;
+			let per_peer_state = self.per_peer_state.read().unwrap();
 
 			for (short_chan_id, mut pending_forwards) in channel_state.forward_htlcs.drain() {
```

Comment:
Hmmm, I'd need some guidance on how to move forward with changing the lock order for the `channel_state` lock here. Due to this line, we'd also need to move the `forward_htlcs` out of the `channel_state` to continue.

Therefore, where would we like to move the `forward_htlcs`? Do we want it in the `peer_state` or would we like it as a _standalone_ lock? Both options have pros and cons.

It a simpler process to move it to a _standalone_ lock, though having it in the `peer_state`, and inserting the htlc to the correct `peer_state` when receiving it could potentially open up for more parallelization when handling it. However the option of storing it in the `peer_state` comes at the cost of having to loop over every `peer_state` and taking the lock in it, to find if any htlcs to be forwarded for that peer. I guess we could combine it with only looping over peers in the `short_to_chan_info` map to not loop over peers with no ready channels though. So given that, I think storing the `forward_htlcs` in the `peer_state` would benefit larger nodes that route a lot of htlcs, while storing it as a standalone lock would benefit smaller nodes.

Which option would we prefer?

Reply:
Thanks! Oh yeah true, that sounds like a recipe for deadlocks if we'd in parallel would process a message from B to A at the same time.

I'll start by moving it into a standalone lock in a separate PR then!